bayes/noise.py

```diff
@@ -1,117 +1,229 @@
+        """
+        raise NotImplementedError("Implement me!")
+
+    def loglike_contribution(self, model_error_dict):
+        """
+        Rearranges the (model_error_key, sensor) ordering of 
+        `model_error_dict` a single numpy matrix and calculates its
+        contribution to a loglikelihood function.
+        """
+        raise NotImplementedError("Implement me!")
+
+
+class UncorrelatedNoiseModel(NoiseModelInterface):
     def __init__(self):
         self.parameter_list = ParameterList()
+        self.parameter_list.define("precision")
+        self._terms = []
+        self._lengths = []  # of the terms for a potential `self.split`
+
+    def add(self, sensor, model_error_key=None):
+        """
+        Adds a (`sensor`, `model_error_key`) pair to the noise model such that 
+        the corresponding output, e.g.
+            model_error_dict[model_error_key][sensor]
+        or 
+            jacobian_dict[model_error_key][sensor]
+        is added to the `self.vector_contribution` or 
+        `self.jacobian_contribution`, respectively.
+        """
+        self._terms.append((sensor, model_error_key))
+
+    def _define_terms(self, model_error_dict):
+        """
+        Can be overwritten to automatically define `self._terms`
+        for certain special cases. See `UncorrelatedSensorNoise` or
+        `UncorrelatedSingleNoise` below.
+        """
+        pass
 
     def vector_contribution(self, model_error_dict):
-        raise NotImplementedError()
+        """
+        overwritten
+        """
+        return self._stack(model_error_dict)
 
-    def loglike_contribution(self, model_error_dict):
-        raise NotImplementedError()
+    def jacobian_contribution(self, jacobian_dict):
+        """
+        overwritten
+        """
+        return self._stack(jacobian_dict)
 
-    def _loglike_term(self, error, sigma):
+    def loglike_contribution(self, model_error_dict):
+        """
+        overwritten
+        """
+        error = self.vector_contribution(model_error_dict)
+        sigma = 1.0 / self.parameter_list["precision"] ** 0.5
         return -0.5 * (
             len(error) * np.log(2.0 * np.pi * sigma ** 2)
             + np.sum(np.square(error / sigma ** 2))
         )
 
+    def _stack(self, dict_of_dicts):
+        """
+        Extracts the (sensor, model_error_key) from the nested `dict_of_dicts`
+        (could be both `model_error_dict` or `jacobian_dict`) and concatenates
+        them to a long numpy vector/matrix.
+
+        It keeps track of the dimension (self._lengths) of the individual terms
+        to reverse this operation in `self.split()`.
+        """
+        self._define_terms(dict_of_dicts)
+        terms = []
+
+        self._lengths = []
+        for (sensor, key) in self._terms:
+            term = dict_of_dicts[key][sensor]
+            self._lengths.append(len(term))
+            terms.append(term)
+        return np.concatenate(terms)
+
+    def split(self, array):
```

Comment:
the naming is clear to me, but not sure for others. Maybe (even though longer) stack -> dict2vector and split -> vector2dict?